File: week-1/problem-set-1/credit.c
#include <cs50.h>
#include <stdio.h>

int main(void)
{
    int card_num_arr[16];
    int remainder, length = 0;
    bool check = false;
    long card_num = get_long("Please, enter the card number: ");
    long card_num_cp = card_num;

    // Adding card_number to an array AND counting its length
    while (card_num_cp > 0)
    {
        remainder = card_num_cp % 10;
        card_num_arr[length] = remainder;
        card_num_cp /= 10;
        length += 1;
    }

    if (length != 16 && length != 13 && length != 15)
    {
      printf("INVALID\n");
      return 0;
    }

    // multiplying each of the underlined digits by 2, starting with the number’s second-to-last digit
    int counter = 1, sum_2nd_to_last = 0, something;
    while (counter < length)
    {
        something = (card_num_arr[counter] * 2) % 10;
        sum_2nd_to_last += ((card_num_arr[counter] * 2) / 10) + something;
        counter += 2;
    }

    int sum_1st_to_last = 0;
    counter = 0;
    while (counter < length)
    {
        sum_1st_to_last += card_num_arr[counter];
        counter += 2;
    }

    // Adding the sum above to the sum of the digits that weren’t multiplied by 2
    int sum = sum_2nd_to_last + sum_1st_to_last;
    if (sum % 10 == 0)
    {
        check = true;
    }

    if (check == true)
    {
        if (card_num_arr[length - 1] == 4)
        {
            printf("VISA\n");
        }
        else if (card_num_arr[length - 1] == 5)
        {
            if (card_num_arr[length - 2] == 1 || card_num_arr[length - 2] == 2 || card_num_arr[length - 2] == 3
                || card_num_arr[length - 2] == 4 || card_num_arr[length - 2] == 5)
            {
                printf("MASTERCARD\n");
            }
            else
            {
                printf("INVALID\n");
            }
        }
        else if (card_num_arr[length - 1] == 3)
        {
            if (card_num_arr[length - 2] == 4 || card_num_arr[length - 2] == 7)
            {
                printf("AMEX\n");
            }
            else
            {
                printf("INVALID\n");
            }
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
}

File: week-6/problem-set-6/sentimental-credit/credit.py
def main():
    # Prompts the user for a credit card number
    card_number = input('Number: ')

    # Determines if a credit card number is not (syntactically) valid
    if not luhns_algorithm(card_number):
        print('INVALID\n')
        exit(1)

    # Determines which company the credit card belongs to
    if not valid_company(card_number):
        print('INVALID\n')
        exit(1)

    exit(0)


def luhns_algorithm(card_num):
    sum = 0

    # Reverse str to simplify the for loop
    card_num = card_num[::-1]

    for i in range(len(card_num)):

        # Multiply every other digit by 2, starting with the number’s second-to-last digit
        if i % 2 != 0:
            product = int(card_num[i]) * 2
            sum += (product // 10) + (product % 10)

        # Add the sum to the sum of the digits that weren’t multiplied by 2
        else:
            sum += int(card_num[i])

    # If the total’s last digit is 0, the number is valid
    return sum % 10 == 0


def valid_company(card_num):
    # AMEX
    if len(card_num) == 15:
        if card_num[:2] == '34' or card_num[:2] == '37':
            print('AMEX\n')
            return True

    # MASTERCARD OR VISA
    elif len(card_num) == 16:

        # MASTERCARD
        if int(card_num[:2]) in range(51, 56):
            print('MASTERCARD\n')
            return True

        # VISA
        elif card_num[0] == '4':
            print('VISA\n')
            return True

    # VISA
    elif len(card_num) == 13:
        if card_num[0] == '4':
            print('VISA\n')
            return True

    return False


main()
